src/ucs/datastruct/callbackq.c

```diff
@@ -7,322 +7,516 @@
+    cbq->fast_elems[idx].cb    = cb;
+    cbq->fast_elems[idx].arg   = arg;
+    cbq->fast_elems[idx].flags = flags;
+    cbq->fast_elems[idx].id    = id;
+    return id;
 }
 
-void ucs_callbackq_cleanup(ucs_callbackq_t *cbq)
+/* should be called from dispatch thread only */
+static void ucs_callbackq_remove_fast(ucs_callbackq_t *cbq, unsigned idx)
 {
-    ucs_callbackq_elem_t *elem;
+    ucs_callbackq_priv_t *priv = ucs_callbackq_priv(cbq);
+    unsigned last_idx;
 
-    /* to execute pending remove requests */
-    ucs_callbackq_invoke_service_cb(cbq);
-    if (cbq->start != cbq->end) {
-        ucs_warn("%zd callbacks still remain in callback queue",
-                 cbq->end - cbq->start);
-        ucs_callbackq_for_each(elem, cbq) {
-            ucs_warn("cbq %p: remain %p %s(arg=%p)", cbq, elem,
-                     ucs_debug_get_symbol_name(elem->cb), elem->arg);
-        }
-    }
-    ucs_free(cbq->ptr);
+    ucs_trace_func("cbq=%p idx=%u", cbq, idx);
+
+    ucs_assert(priv->num_fast_elems > 0);
+    last_idx = --priv->num_fast_elems;
+    ucs_callbackq_remove_common(cbq, cbq->fast_elems, idx, last_idx, 0,
+                                &priv->fast_remove_mask);
 }
 
-void ucs_callbackq_add(ucs_callbackq_t *cbq, ucs_callback_t cb, void *arg)
+/* should be called from dispatch thread only */
+static void ucs_callbackq_purge_fast(ucs_callbackq_t *cbq)
 {
-    ucs_callbackq_elem_t *elem;
+    ucs_callbackq_priv_t *priv = ucs_callbackq_priv(cbq);
+    unsigned idx;
 
-    ucs_callbackq_enter(cbq);
+    ucs_trace_func("cbq=%p map=0x%"PRIx64, cbq, priv->fast_remove_mask);
 
-    elem = ucs_callbackq_find(cbq, cb, arg);
-    if (elem != NULL) {
-        ucs_atomic_add32(&elem->refcount, 1);
-        ucs_callbackq_leave(cbq);
-        return;
+    /* Remove fast-path callbacks marked for removal */
+    while (priv->fast_remove_mask) {
+        idx = ucs_ffs64(priv->fast_remove_mask);
+        ucs_callbackq_remove_fast(cbq, idx);
     }
+}
 
-    if (cbq->end >= cbq->ptr + cbq->size) {
-        /* TODO support expanding the callback queue */
-        ucs_fatal("callback queue %p is full, cannot add %s()", cbq,
-                  ucs_debug_get_symbol_name(cb));
-    }
+static void ucs_callbackq_enable_proxy(ucs_callbackq_t *cbq)
+{
+    ucs_callbackq_priv_t *priv = ucs_callbackq_priv(cbq);
+    unsigned idx;
+    int id;
 
-    elem = cbq->end;
+    ucs_trace_func("cbq=%p", cbq);
 
-    ucs_trace("cbq %p: adding %p %s(arg=%p) [start:%p end:%p]", cbq, elem,
-              ucs_debug_get_symbol_name(cb), arg, cbq->start, cbq->end);
+    if (priv->slow_proxy_id != UCS_CALLBACKQ_ID_NULL) {
+        return;
+    }
 
-    elem->cb       = cb;
-    elem->arg      = arg;
-    elem->refcount = 1;
+    ucs_assert((priv->num_slow_elems > 0) || priv->fast_remove_mask);
 
-    /* Make sure a thread dispatching the callbacks would see 'end' only after
-     * the new element is set.
+    idx = ucs_callbackq_get_fast_idx(cbq);
```

Comment:
This (ucs_callbackq_enable_proxy) function does not check it. And if you already have 7 FP callbacks, enabling slow-path proxy will fail with assertion. Assertion checks for value to be not more than UCS_CALLBACKQ_FAST_MAX (rather than UCS_CALLBACKQ_FAST_COUNT )